contracts/ens/ens.go

```diff
@@ -181,15 +230,71 @@ func (ens *ENS) Register(name string) (*types.Transaction, error) {
 }
 
 // SetContentHash sets the content hash associated with a name. Only works if the caller
-// owns the name, and the associated resolver implements a `setContent` function.
-func (ens *ENS) SetContentHash(name string, hash common.Hash) (*types.Transaction, error) {
+// owns the name, and the associated resolver implements a `setContenthash` function.
+func (ens *ENS) SetContentHash(name string, hash []byte) (*types.Transaction, error) {
 	node := EnsNode(name)
 
 	resolver, err := ens.getResolver(node)
 	if err != nil {
 		return nil, err
 	}
+
 	opts := ens.TransactOpts
 	opts.GasLimit = 200000
-	return resolver.Contract.SetContent(&opts, node, hash)
+
+	// IMPORTANT: The old contract is deprecated. This code should be removed latest on June 1st 2019
+	supported, err := resolver.SupportsInterface(contentHash_Interface_Id)
+	if err != nil {
+		return nil, err
+	}
+
+	if !supported {
+		resolver, err := ens.getFallbackResolver(node)
+		if err != nil {
+			return nil, err
+		}
+		opts := ens.TransactOpts
+		opts.GasLimit = 200000
+		var b [32]byte
+		copy(b[:], hash)
+		return resolver.Contract.SetContent(&opts, node, b)
+	}
+
+	// END DEPRECATED CODE
+	return resolver.Contract.SetContenthash(&opts, node, hash)
+}
+
+func decodeMultiCodec(b []byte) (common.Hash, error) {
+
+	// Create a cid from a marshaled string
+	_, err := cid.Decode(string(b))
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	return common.Hash{}, nil
+	/* from the EIP documentation
+	   storage system: Swarm (0xe4)
+	   CID version: 1 (0x01)
+	   content type: swarm-manifest (0x??)
```

Comment:
I think you should be using a multicodec content type from this list: https://ipfs.io/ipfs/QmXec1jjwzxWJoNbxQF5KffL8q6hFXm9QwUGaa3wKGk6dT/#title=Multicodecs&src=https://raw.githubusercontent.com/multiformats/multicodec/master/table.csv

From the documentation:
```
<multicodec-content-type> is a multicodec code representing the content type or format of the data being addressed.
```

https://github.com/multiformats/multicodec

If we need a content type that is not available, we must submit a PR to `multicodec` repo.

Reply:
https://github.com/multiformats/multicodec/pull/120

Reply:
@nonsense, I'm actually not 100% sure that I want to vendor so much code just to use the CID functionality.
On the flip-side, vendoring it means easier handling of keeping up to date with changes to the multicodec specs. On the third side - their code repos and go libraries do not seem to have any code generators in place that will automatically define the types according to what's defined in the `multicodec` table on `master`. Which essentially means - if we change the table with the swarm type codes we'd still need to PR it to the `go-multihash` library for example :/